azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -48,3 +63,331 @@ def _execute_op(self, op):
+    # def _get_retry_value(self, provisioning_op):
+    #     """
+    #     Return None if there is an error or else return the retry-after value obtained from the
+    #     service response. If the service has not given a retry-afer then return a default value.
+    #     The only reason a operation will not be retried is if it has moved to a different operation or else it has completed successfully.
+    #     In this case it means that a SendRegistrationRequestOperation
+    #     """
+    #     key_values_dict = provisioning_op.key_values
+    #     retry_after = (
+    #         None if "retry-after" not in key_values_dict else str(key_values_dict["retry-after"][0])
+    #     )
+    #     return retry_after
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _process_error(self, provisioning_op, prov_op_name, error):
+        if error:
+            return error
+        elif 300 <= provisioning_op.status_code < 429:
+            logger.error(
+                "Received error with status code {status_code} for {prov_op_name} request operation".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+            logger.error("response body: {}".format(provisioning_op.response_body))
+            return exceptions.ServiceError(
+                "{prov_op_name} request returned a service error status code {status_code}".format(
+                    prov_op_name=prov_op_name, status_code=provisioning_op.status_code
+                )
+            )
+        else:
+            return None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _decode_response(self, provisioning_op):
+        return json.loads(provisioning_op.response_body.decode("utf-8"))
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_registration_status(self, decoded_response):
+
+        status = None if "status" not in decoded_response else str(decoded_response["status"])
+        return status
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _get_operation_id(self, decoded_response):
+
+        operation_id = (
+            None if "operationId" not in decoded_response else str(decoded_response["operationId"])
+        )
+        return operation_id
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _form_complete_result(self, operation_id, decoded_response, status):
+        """
+        Create the registration result from the complete decoded json response for details regarding the registration process.
+        """
+        decoded_state = (
+            None
+            if "registrationState" not in decoded_response
+            else decoded_response["registrationState"]
+        )
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
+            registration_state = RegistrationState(
```

Comment:
>registration_state [](start = 12, length = 18)

(I hate how our pretty-printer works sometimes.  It does not make this code readable.)
suggestion: I see this pattern a bunch.  How about something like this: device_id=get_optional_element(decoded_state, "deviceId") where get_optional_element is your helper that returns None if the element is missing from the dict?

Reply:
changed to a helper function defined in `mqtt_topic.py`. I feel that should be renamed to `mqtt_helper.py` since it is extracting stuff from topic as well as response...
The thing from which the optional element could be retrieved could be anything either string or dictionary

---
In reply to: [345906762](https://github.com/Azure/azure-iot-sdk-python/pull/391#discussion_r345906762) [](ancestors = 345906762)